### PR TITLE
Implement separate functions for get_device from id and array

### LINKED ISSUE #2051: Implement separate functions for get_device from id and array
cuda.get_device accepts both a device id and an object. When an id is given, it returns a device of the corresponding ordinal. When an object is given, it checks the type and returns a device if the type is cupy.ndarray. This behavior is complicated and often introduces bugs when we take care of the performance. I think it is enough to provide two functions: get_device_from_id (which takes an id and returns the corresponding device) and get_device_from_array (which takes an input and returns the device if the argument is cupy.ndarray). It much simplifies the interface and implementation.

### PR DESCRIPTION
**Pull request by mitmul**
Fix #2051

### CODE REVIEW

File: chainer/cuda.py [hunk 1]
```diff
@@ -141,41 +141,51 @@ def __ne__(self, other):
...
 
     Args:
-        args: Values to specify a GPU device. The first device object, integer
-            or :class:`cupy.ndarray` object is used to select a device.
-            If it is a device object, it is returned. If it is an integer,
-            the corresponding device is returned. If it is a CuPy array,
-            the device on which this array reside is returned. If any
-            arguments are neither integers nor CuPy arrays, a dummy device
-            object representing CPU is returned.
+        device_id (int): The ID of the device which this function returns.
```
Comment (unnonouno) [inline chainer/cuda.py]: Which type of argument do we need to accept except integers? I think we need to specify constraint here.

File: chainer/cuda.py [hunk 2]
```diff
@@ -141,41 +141,51 @@ def __ne__(self, other):
...
-            return arg
+            return array.device
+    return DummyDevice
+
+
+def get_device(device):
+    """Checks and gets the device regardless of it's None or not.
 
+    Args:
+        device (:class:`cupy.ndarray` or None): If `None`, it returns
```
Comment (unnonouno) [inline chainer/cuda.py]: `cuda.Device`?

File: chainer/functions/loss/crf1d.py [hunk 3]
```diff
@@ -122,9 +122,8 @@ def crf1d(cost, xs, ys):
             y_prev, _ = split_axis.split_axis(y_prev, [batch], axis=0)
             score, score_rest = split_axis.split_axis(score, [batch], axis=0)
             scores.append(score_rest)
-        score += (select_item.select_item(x, y)
-                  + reshape.reshape(
-                      embed_id.embed_id(y_prev * n_label + y, cost), (batch,)))
+        score += (select_item.select_item(x, y) + reshape.reshape(
```
Comment (unnonouno) [inline chainer/functions/loss/crf1d.py]: Is this fix required?

File: docs/source/reference/util/cuda.rst [hunk 4]
```diff
@@ -4,7 +4,8 @@ CUDA utilities
 
 Devices
 ~~~~~~~
-.. autofunction:: get_device
+.. autofunction:: get_device_from_id
+.. autofunction:: get_device_from_array
```
Comment (unnonouno) [inline docs/source/reference/util/cuda.rst]: Is get_device deprecated?

File: chainer/cuda.py (hunk 1 above)
Comment (mitmul) [inline chainer/cuda.py]: It only accepts integers. So I'll add an exception processing here.

Comment (mitmul) [inline chainer/cuda.py]: Hmm, `get_device_from_id` sometimes should be able to take `None` as the input
 `device` argument. So simple exception processing doesn't work.

Comment (unnonouno) [inline chainer/cuda.py]: Is ``None`` acceptable here?

Comment (unnonouno) [inline chainer/cuda.py]: How about `(int or None)` and writing behavior when `None` is given?
And i think this code is enough:
```
if device_id is None:
    return DummyDevice
else:
    return Device(device_id)
```

**Comment by unnonouno** [comment]
LGTM. Please merge _v2 and resolve conflicts.

**Comment by unnonouno** [comment]
And I think we can merge this feature in v1.

**Comment by mitmul** [comment]
I noticed that I can remove `get_device` method completely from Chainer now. I'm fixing this PR.

**Comment by mitmul** [comment]
@unnonouno As for this comment,
> And I think we can merge this feature in v1.
Finally I removed `get_device` method completely. So it's not possible now...

**Comment by unnonouno** [comment]
I mean, we can add new methods and replace old one in v1, make old API deprecated and remove old API in v2.

**Comment by mitmul** [comment]
OK, I understand. But it breaks backward compatibility, so I think it's better to merge this only into v2. Removing `get_device` also from v1 sounds a bit risky. How do you think on it? @beam2d

**Comment by beam2d** [comment]
We should not remove `get_device` from v1. As @unnonouno wrote, we have to leave it at least with deprecation warning. How about making this PR only deprecate `get_device` and add `get_device_from_*`, merge it into master, and make another PR targetting v2 that just removes `get_device`?

**Comment by mitmul** [comment]
OK, I'll make other PRs. So I close this. Thanks

**Comment by mitmul** [comment]
Sorry, applying this after #2391 can achieve what you guys thought? So I re-opened it.

**Comment by unnonouno** [comment]
We need to change `get_device` with newer APIs in v1. But does the newer API support `Device` object?

File: docs/source/tutorial/gpu.rst [hunk 5]
```diff
@@ -107,29 +107,30 @@ It is equivalent to the following code using CuPy:
    If user uses only one device, these device switching is not needed.
    :func:`chainer.cuda.to_cpu` and :func:`chainer.cuda.to_gpu` functions automatically switch the current device correctly.
 
-Chainer also provides a convenient function :func:`chainer.cuda.get_device` to select a device.
-It accepts an integer, CuPy array, NumPy array, or None (indicating the current device), and returns an appropriate device object.
-If the argument is a NumPy array, then *a dummy device object* is returned.
+Chainer also provides convenient functions :func:`chainer.cuda.get_device_from_id` and :func:`chainer.cuda.get_device_from_array` to select a device.
+The former one accepts an integer indicating the target device and returns an appropriate device object.
+The latter one accepts a :class:`cuda.ndarray` object and returns the corresponding device object.
+In this latter function, if the argument is a NumPy array, then *a dummy device object* is returned.
```
Comment (niboshi) [inline docs/source/tutorial/gpu.rst]: * I think the word "convenient" is no longer applicable. The word implies doing something comlicated for the user, but its functionalities are now simplified.
* Each appearance of the word "one" seems redundant for me. "The former[latter] accepts" feels like more natural.

File: tests/chainer_tests/test_cuda.py [hunk 6]
```diff
@@ -32,16 +32,14 @@ def test_ne(self):
...
 class TestCuda(unittest.TestCase):
 
-    def test_get_dummy_device(self):
-        self.assertIs(cuda.get_device(), cuda.DummyDevice)
-
-    def test_get_device_for_numpy_int(self):
-        self.assertIs(cuda.get_device(numpy.int64(0)), cuda.DummyDevice)
+    def test_get_device_from_id_for_numpy_int(self):
+        self.assertIs(cuda.get_device_from_id(numpy.int64(0)),
+                      cuda.DummyDevice)
```
Comment (niboshi) [inline tests/chainer_tests/test_cuda.py]: Should it be `cuda.get_device_from_array`?
This is out of specification of  `cuda.get_device_from_id`. It should throw an error.

File: chainer/link.py [hunk 7]
```diff
@@ -265,16 +265,16 @@ def to_cpu(self):
...
 
         This method does not handle non-registered attributes. If some of such
         attributes must be copied to GPU, the link implementation must
         override this method to do so.
 
         Args:
-            device: Target device specifier. If omitted, the current device is
-                used.
+            device_id (int): The device id to specify the target device. If
+                omitted, the current device is used.
```
Comment (niboshi) [inline chainer/link.py]: The expression should match the one in `variable.Variable` documentation.

**Pull request closed without merging**